### PR TITLE
API: Rescale the object and probe together

### LINKED ISSUE #47: Normalize the object array to 1 on average
> In order to avoid amplitude ambiguity in psi and keep it 1 on average, on each iteration we can compute |FQψ| and compare its norm with the norm of data. Then compute the constant c = norm(|FQψ) / norm(sqrt(data)) and do Q/=c. So this way we simply replace prb by prb/c and ψ by ψ*c, which does not affect the functional |FQ/c ψ*c| = |FQψ|). @nikitinvv

### PR DESCRIPTION
**Pull request by carterbox**
## Purpose

Make comparison of different ptychography reconstruction algorithms easier by normalizing the probe and object result. 

Closes #47 
## Approach

The probe and object are scaled by a constant every iteration. The constant is the average amplitude of the object; dividing the object by the average amplitude keeps the average amplitude around 1.

### Submitter
- [x] Write a helpfully descriptive pull request title.
- [x] Organize changes into logically grouped commits with descriptive commit messages.
- [ ] Document all new functions.
- [ ] Write tests for new functions or explain why they are not needed.
- [ ] Build the documentation successfully
- [ ] Use [`yapf`](https://github.com/google/yapf) to format python code.

### Reviewer
- [ ] Actually read all of the code.
- [ ] Run the new code yourself.
- [ ] Write a summary of the changes as you understand them.
- [ ] Thank the submitter.

### CODE REVIEW

**Comment by nikitinvv** [comment]
This scaling should be done on each iteration, keeping np.mean(psi)~1 if the initial guess for psi was 1. With this, we will get a uniform background in the amplitude of psi.  It does not make sense to do it after all iterations since it will not affect the background.

**Comment by carterbox** [comment]
When I switched to updating every iteration, I observed that the rescaling was not working. The rescaling value was the same every iteration which caused the values of the object and probe to diverge.

**Comment by carterbox** [comment]
I will fix the tests after we have agreed that the rescaling is correct.

**Comment by nikitinvv** [comment]
normalization constant is not computed by using psi. For some constant c we want to have  |F(cQ^i)ψ| of the order of data_i, where i is a scan position. We cannot do this for all i separately. So we can consider norms and find c s.t. \norm(|F(cQ)\psi|) is of the order of \norm(data). From there we see that c = \norm(data) / \norm(FQ\psi). After computing c we incorporate it in the probe and do Q*=c. This way we fit the background: |FQ^i 1| \approx data_i for i in background

File: src/tike/ptycho/ptycho.py [hunk 1]
```diff
@@ -192,7 +192,25 @@ def reconstruct(
...
         raise ValueError(
             "The '{}' algorithm is not an available.".format(algorithm))
+
+
+def _rescale_obj_probe(operator, data, psi, scan, probe):
+    """Keep the object amplitude around 1 by scaling by a constant"""
+    intensity = operator._compute_intensity(data, psi, scan, probe)
+
+    rescale = (np.linalg.norm(np.ravel(np.sqrt(intensity))) /
+               np.linalg.norm(np.ravel(np.sqrt(data))))
```
Comment (carterbox) [inline src/tike/ptycho/ptycho.py]: Yes, I tried the method you proposed in #47. If the objective is to keep `np.mean(np.abs(psi))` near `1`, then it doesn't work. Instead the amplitude of the object slowly creeps toward zero and the probe amplitude moves to infinity.

**Comment by carterbox** [comment]
Yes, I tried the method you proposed in #47 in 346dcca If the objective is to keep `np.mean(np.abs(psi))` near `1`, then it doesn't work. Instead the amplitude of the object slowly creeps toward zero and the probe amplitude moves to infinity.

**Comment by nikitinvv** [comment]
On each iteration do this: c = \norm(data) / \norm(FQ\psi); Q*=c. Dont do this psi *= c

**Comment by carterbox** [comment]
Only rescaling the probe doesn't work. The `norm / norm` rescaling value never reaches 1, so the probe intensity slows moves to infinity.

**Pull request merged**